executor/adapter.go

```diff
@@ -1504,84 +1505,97 @@ func resetCTEStorageMap(se sessionctx.Context) error {
-		traverseFlatPlanForHistoryStats(stmtCtx, flatSubQueryPlanTree, sql)
+		succ = traverseFlatPlanForHistoryStats(stmtCtx, statsHandle, startTS, flatSubQueryPlanTree)
+		if !succ {
+			return
+		}
 	}
 }
 
 func isTableFullScanWithoutFilters(tableScan *plannercore.PhysicalTableScan) bool {
 	return tableScan.HasRFFilters() || len(tableScan.AccessCondition) > 0 || tableScan.HasLateMaterializationFilterCondition()
 }
 
-// traverseFlatPlanForHistoryStats returns true when it contains limit operator, because limit operator may lead to early
-// exit, and make stats info inaccurate.
-func traverseFlatPlanForHistoryStats(stmtCtx *stmtctx.StatementContext, flatPlanTree plannercore.FlatPlanTree, sql string) bool {
+// traverseFlatPlanForHistoryStats returns false when
+// 1. Plan contains limit operator, because limit operator may lead to early exit
+// 2. Fails to update history stats
+func traverseFlatPlanForHistoryStats(stmtCtx *stmtctx.StatementContext, statsHandle *handle.Handle, startTS uint64, flatPlanTree plannercore.FlatPlanTree) bool {
 	for _, flatOp := range flatPlanTree {
 		pp, isPhysicalPlan := flatOp.Origin.(plannercore.PhysicalPlan)
 		if !isPhysicalPlan {
 			continue
 		}
 		switch x := pp.(type) {
 		case *plannercore.PhysicalLimit:
-			return true
+			return false
 		case *plannercore.PhysicalSelection:
 			if tableScan, ok := x.Children()[0].(*plannercore.PhysicalTableScan); ok {
 				if isTableFullScanWithoutFilters(tableScan) {
 					break
 				}
-				needRecord, estRows, actualRows := needRecordHistoryStats(stmtCtx, x)
+				needRecord, _, actualRows := needRecordHistoryStats(stmtCtx, x)
 				if !needRecord {
 					break
 				}
+				hasCopStats, totalTableRows := GetResultRowsCountForCopStats(stmtCtx, tableScan.ID())
+				if !hasCopStats {
+					break
+				}
+				totalTableRowsFloat := float64(totalTableRows)
 				succ, hashCode := genHistoryStatsHashCodeForSelection(stmtCtx, x)
 				if !succ {
 					break
 				}
-				recordHistoryStats(sql, tableScan.Table.ID, x.ID(), false, hashCode, estRows, actualRows)
-			}
-		case *plannercore.PhysicalTableScan:
-			if isTableFullScanWithoutFilters(x) {
-				break
-			}
-			needRecord, estRows, actualRows := needRecordHistoryStats(stmtCtx, x)
-			if !needRecord {
-				break
+				err := recordHistoryStats(statsHandle, startTS, tableScan.Table.ID, hashCode, actualRows, totalTableRowsFloat)
+				if err != nil {
+					logutil.BgLogger().Warn(err.Error())
+					return false
+				}
 			}
-			recordHistoryStats(sql, x.Table.ID, x.ID(), true, 0, estRows, actualRows)
 		}
 	}
-	return false
+	return true
 }
 
-func recordHistoryStats(sql string, tableID int64, planID int, nullHashCode bool, hashCode uint64, estRows float64, actualRows int64) {
+func recordHistoryStats(statsHandle *handle.Handle, startTS uint64, tableID int64, hashCode uint64, actualRows int64, totalTableRows float64) error {
+	var selectivity float32 = 0.0
+	if totalTableRows != 0.0 {
```

Comment:
Can you exit if totalTableRows == 0.0? This case is not interesting to capture.

Reply:
Will change it.